src/analysis/normed_space/operator_norm.lean

```diff
@@ -0,0 +1,217 @@
+lemma exists_bound' (A : L(E,F)) : ∃ c, c > 0 ∧ ∀ x : E, ∥x∥ ≤ 1 → ∥A x∥ ≤ c :=
+let ⟨c, _, H⟩ := exists_bound A in
+exists.intro c ⟨‹c > 0›,
+  assume x _,
+  calc ∥A x∥ ≤ c * ∥x∥ : H x
+        ... ≤ c * 1 : (mul_le_mul_left ‹c > 0›).mpr ‹∥x∥ ≤ 1›
+        ... = c : mul_one c⟩
+
+end bounded_linear_maps
+
+/-
+Now define the operator norm. We only do this for normed spaces over ℝ, since we need a
+scalar multiplication with reals to prove that ∥A x∥ ≤ ∥A∥ * ∥x∥. It would be enough to
+have a vector space over a normed field k with a real scalar multiplication and certain
+compatibility conditions.
+
+The main task is to show that the operator norm is definite, homogeneous, and satisfies the
+triangle inequality. This is done after a few preliminary lemmas necessary to deal with cSup.
+-/
+section operator_norm
+
+variables [normed_space ℝ E] [normed_space ℝ F]
+open lattice set
+
+local notation `L(` E `,` F `)` := @bounded_linear_maps _ E F _ _ _
+
+/-- The operator norm of a bounded linear map A : E → F is the sup of
+    the set ∥A x∥ with ∥x∥ ≤ 1. If E = {0} we set ∥A∥ = 0. -/
+noncomputable def operator_norm (A : L(E, F)) : ℝ :=
+Sup (image (λ x, ∥A x∥) {x | ∥x∥ ≤ 1})
+
+noncomputable instance bounded_linear_maps.to_has_norm : has_norm L(E,F) :=
+{norm := operator_norm}
+
+lemma norm_of_unit_ball_bdd_above (A : L(E,F)) : bdd_above (image (norm ∘ A) {x | ∥x∥ ≤ 1}) :=
+let ⟨c, _, H⟩ := (exists_bound' A : ∃ c, c > 0 ∧ ∀ x : E, ∥x∥ ≤ 1 → ∥A x∥ ≤ c) in
+bdd_above.mk c
+  (assume r ⟨x, (_ : ∥x∥ ≤ 1), (_ : ∥A x∥ = r)⟩,
+    show r ≤ c, from 
+      calc r = ∥A x∥ : eq.symm ‹∥A x∥ = r›
+         ... ≤ c : H x ‹∥x∥ ≤ 1›)
+
+lemma zero_in_im_ball (A : L(E,F)) : (0:ℝ) ∈ {r : ℝ | ∃ (x : E), ∥x∥ ≤ 1 ∧ ∥A x∥ = r} :=
+have A 0 = 0, from linear_map.map_zero {to_fun := A.val, ..A.property}, 
+exists.intro (0:E) $ and.intro (by rw[norm_zero]; exact zero_le_one) (by rw[‹A 0 = 0›]; simp)
+
+lemma operator_norm_nonneg (A : L(E,F)) : 0 ≤ ∥A∥ :=
+have (0:ℝ) ∈ _, from zero_in_im_ball A,
+suffices 0 ≤ Sup (image (norm ∘ A) {x | ∥x∥ ≤ 1}), by assumption,
+let ⟨c, _, H⟩ := (exists_bound' A : ∃ c, c > 0 ∧ ∀ x : E, ∥x∥ ≤ 1 → ∥A x∥ ≤ c) in
+le_cSup (norm_of_unit_ball_bdd_above A) ‹(0:ℝ) ∈ _›
+
+lemma bounded_by_operator_norm_on_unit_vector (A : L(E, F)) {x : E} (_ : ∥x∥ = 1) : ∥A x∥ ≤ ∥A∥ :=
+suffices ∥A x∥ ≤ Sup (image (norm ∘ A) {x | ∥x∥ ≤ 1}), by assumption,
+let ⟨c, _, _⟩ := (exists_bound A : ∃ c, c > 0 ∧ ∀ x : E, ∥ A x ∥ ≤ c * ∥ x ∥) in
+have ∥A x∥ ∈ (image (norm ∘ A) {x | ∥x∥ ≤ 1}), from exists.intro x ⟨le_of_eq ‹∥x∥ = 1›, rfl⟩,
+le_cSup (norm_of_unit_ball_bdd_above A) ‹∥A x∥ ∈ _›
+
+/-- This is the fundamental property of the operator norm: ∥A x∥ ≤ ∥A∥ * ∥x∥. -/
+theorem bounded_by_operator_norm {A : L(E,F)} {x : E} : ∥A x∥ ≤ ∥A∥ * ∥x∥ :=
+have A 0 = 0, from linear_map.map_zero {to_fun := A.val, ..A.property}, 
+classical.by_cases
+  (assume : x = (0:E),
+    calc ∥A x∥ ≤ 0 : by rw[‹x = 0›, ‹A 0 = 0›, norm_zero]; exact le_refl 0
```

Comment:
This can be solved in one step: `show ∥A x∥ ≤ ∥A∥ * ∥x∥, by rw [‹x = 0›, ‹A 0 = 0›, norm_zero, mul_zero]; exact le_refl 0`